src/v/iceberg/compatibility.h

```diff
@@ -0,0 +1,29 @@
+// Copyright 2024 Redpanda Data, Inc.
+//
+// Use of this software is governed by the Business Source License
+// included in the file licenses/BSL.md
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0
+
+#pragma once
+
+#include "base/outcome.h"
+#include "iceberg/datatypes.h"
+
+#include <seastar/util/bool_class.hh>
+
+namespace iceberg {
+
+enum class compat_errc {
+    mismatch,
+};
+
+using type_promoted = ss::bool_class<struct type_promoted_tag>;
+using type_check_result = checked<type_promoted, compat_errc>;
+
+type_check_result
+check_types(const iceberg::field_type& src, const iceberg::field_type& dest);
```

Comment:
nit: some light docs would be 😎

Reply:
+1, it'd be great to call out here that this is primarily checking the primitives and isn't expected to be used as a standalone nested type-promotion checker

Reply:
done